app/liquid_tags/user_subscription_tag.rb

```diff
@@ -1,9 +1,242 @@
+              source_type: "Article",
+              source_id: articleId,
+              subscriber_email: subscriber.email
+            }
+          }
+        )
+
+      return fetch('/user_subscriptions', {
+        method: 'POST',
+        headers: headers,
+        credentials: 'same-origin',
+        body: body,
+      }).then(function(response) {
+        return response.json();
+      });
+    }
+
+    function fetchIsSubscribed() {
+      const articleId = document.getElementById('article-body').dataset.articleId;
+
+      const params = new URLSearchParams({
+        source_type: 'Article',
+        source_id: articleId
+      }).toString();
+
+      const headers = {
+        Accept: 'application/json',
+        'X-CSRF-Token': window.csrfToken,
+        'Content-Type': 'application/json',
+      }
+
+      return fetch(`/user_subscriptions/subscribed?${params}`, {
+        method: 'GET',
+        headers: headers,
+        credentials: 'same-origin',
+      }).then(function(response) {
+        if (response.ok) {
+          return response.json();
+        } else {
+          console.error(`Base data error: ${response.status} - ${response.statusText}`);
+        }
+      });
+    }
+
+    // Handle API responses
+    // ***************************************
+    function handleSubscription() {
+      submitSubscription().then(function(response) {
+        if (response.success) {
+          const authorUsername = document.getElementById('user-subscription-tag').dataset.authorUsername;
+          const successMsg = `You are now subscribed and may receive emails from ${authorUsername}`;
+          showResponseMessage('success', successMsg);
+        } else {
+          showResponseMessage('danger', response.error);
+        }
+      });
+    }
+
+    function checkIfSubscribed() {
+      fetchIsSubscribed().then(function(response) {
+        const subscriber = userData();
+        const isSubscriberAuthedWithApple = subscriber.email.endsWith('@privaterelay.appleid.com');
+
+        if (response.is_subscribed) {
+          showSubscribed();
+        } else if (isSubscriberAuthedWithApple) {
+          showAppleAuthMessage();
+        } else {
+          updateSubscriberData();
+        }
+      });
+    }
+
+    // The markup defaults to signed out UX
+    if (isUserSignedIn()) {
+      showSignedIn();
+      addConfirmationModalClickHandlers();
+
+      // We need access to some DOM elements (i.e. csrf token, article id, userData, etc.)
+      document.addEventListener("DOMContentLoaded", function() {
+        checkIfSubscribed();
+      });
+    } else {
+      addSignInClickHandler();
+    }
+  JAVASCRIPT
+
   def initialize(_tag_name, cta_text, _tokens)
     @cta_text = cta_text.strip
   end
 
-  def render(context); end
+  def render(context)
+    source = detect_in_context(:source, context)
+    author = source&.user
+    author_profile_image = author&.profile_image_90
+    author_username = author&.username
```

Comment:
`source` will be `nil` in some circumstances (i.e. Article editor preview). And that's okay, we'll just pass `nil` values to the view and those items just show up "blank" 👍